File: workoutpixel/src/main/java/ch/karimattia/workoutpixel/database/GoalRepository.java
package ch.karimattia.workoutpixel.database;

import android.app.Application;

import androidx.lifecycle.LiveData;

import java.util.List;

import ch.karimattia.workoutpixel.core.Goal;

class GoalRepository {

    private GoalDao goalDao;
    private LiveData<List<Goal>> allGoals;

    // Note that in order to unit test the WordRepository, you have to remove the Application
    // dependency. This adds complexity and much more code, and this sample is not about testing.
    // See the BasicSample in the android-architecture-components repository at
    // https://github.com/googlesamples
    GoalRepository(Application application) {
        AppDatabase db = AppDatabase.getDatabase(application);
        goalDao = db.goalDao();
        allGoals = goalDao.loadAllGoalsLiveData();
    }

    // Room executes all queries on a separate thread.
    // Observed LiveData will notify the observer when the data has changed.
    LiveData<List<Goal>> getAllGoals() {
        return allGoals;
    }

/*
    // You must call this on a non-UI thread or your app will throw an exception. Room ensures
    // that you're not doing any long running operations on the main thread, blocking the UI.
    void insert(Word word) {
        WordRoomDatabase.databaseWriteExecutor.execute(() -> {
            mWordDao.insert(word);
        });
    }
*/
}

File: app/src/main/java/ch/karimattia/workoutpixel/data/GoalRepository.kt
package ch.karimattia.workoutpixel.data

import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

@Suppress("unused")
private const val TAG = "GoalRepository"

class GoalRepository @Inject constructor(
	private val goalDao: GoalDao,
) {
	// Room executes all queries on a separate thread.
	// Observed Flow will notify the observer when the data has changed.

	val allGoals: Flow<List<Goal>> = goalDao.loadAllGoalsFlow()

	suspend fun deleteGoal(goal: Goal) = goalDao.deleteGoal(goal = goal.goalWithoutCount())
	suspend fun insertGoal(goal: Goal): Int =
		goalDao.insertGoal(goal = goal.goalWithoutCount()).toInt()

	suspend fun updateGoal(goal: Goal) = goalDao.updateGoal(goal = goal.goalWithoutCount())
	suspend fun loadGoalByUid(uid: Int): Goal? = goalDao.loadGoalByUid(uid = uid)
	fun loadGoalByAppWidgetIdFlow(appWidgetId: Int): Flow<Goal> =
		goalDao.loadGoalByAppWidgetIdFlow(appWidgetId)

	fun loadGoalsWithoutValidAppWidgetId(): Flow<List<Goal>> =
		goalDao.loadGoalsWithoutValidAppWidgetId()

	suspend fun loadGoalsWithValidAppWidgetId(): List<Goal> =
		goalDao.loadGoalsWithValidAppWidgetId()

	suspend fun setAppWidgetIdToNullByAppwidgetId(appWidgetId: Int) =
		goalDao.setAppWidgetIdToNullByAppwidgetId(appWidgetId = appWidgetId)

	suspend fun loadGoalByAppWidgetId(appWidgetId: Int): Goal? =
		goalDao.loadGoalByAppWidgetId(appWidgetId = appWidgetId)

	suspend fun setAppWidgetIdToNullByUid(uid: Int) = goalDao.setAppWidgetIdToNullByUid(uid = uid)

	suspend fun allGoals(): List<Goal> = goalDao.loadAllGoals()
}